Rubberduck.Parsing/Symbols/DeclarationFinder.cs

```diff
@@ -688,15 +688,21 @@ public Declaration FindMemberEnclosingModule(Declaration callingModule, Declarat
 
         public Declaration FindMemberEnclosingProcedure(Declaration enclosingProcedure, string memberName, DeclarationType memberType, ParserRuleContext onSiteContext = null)
         {
-            if (memberType == DeclarationType.Variable && NameComparer.Equals(enclosingProcedure.IdentifierName, memberName))
-            {
-                return enclosingProcedure;
-            }
             var allMatches = MatchName(memberName);
             var memberMatches = allMatches.Where(m =>
                 m.DeclarationType.HasFlag(memberType)
                 && enclosingProcedure.Equals(m.ParentDeclaration));
-            return memberMatches.FirstOrDefault();
+            if (memberMatches.Any())
+            {
+                return memberMatches.FirstOrDefault();
+            }
+
+            if (memberType == DeclarationType.Variable && NameComparer.Equals(enclosingProcedure.IdentifierName, memberName))
+            {
+                return enclosingProcedure;
+            }
+
```

Comment:
It looks like reversing the order of the two `if` clauses does not affect the method outcome, and the code might avoid the call to `MatchName(memberName)` .

Reply:
AFAIK, this function is used to find member decalartion both in procedures and functions. 
In functions the new order is actually a bit more problematic than the old one. However, both the new and the old the logic are wrong anyway: the result must depend on the context of the member and the type of the enclosing procedure. (This causes another bug in RD whose issue(s) I could not find with my current level of search-fu.)  

In a function with a parameter with the same name, on the LHS of an assignment, the name of an enclosing function has to resolve to the function; on the RHS, it has to resolve to a parameter with the same name. With the new order, it will always resolve to the parameter, never to the function.

Please note that the order would be correct for procedures. So, the logic has to be enhanced a bit to both depend on the context and the type of the enclosing procedure.

Reply:
Found the issue #3499.

Reply:
In next sync, I will include the new three unit tests I added to account for the recursive calls as I could not find any existing unit tests that covered this case to guard against this regression. As far as I can tell, they pass.

Reply:
Thanks; will add unit tests for the #3499 case and see what can be done.

Reply:
@MDoerner  just to follow up on your comments....

I've determined that it is not possible to create a function that also has a parameter of the same name; this would be considered a compile-time error by VBA. It is only possible to create a `Sub` or property procedures that can have parameter with the same name as the routine's name and we have unit tests to cover those cases. For this reason, having parameters taking priority over variables seems to be most reasonable to do in this case.

Reply:
My previous comment was based on the discussion in the issue referenced. I tested myself now and have to say that you are completely right. In property set and let methods, the reference is always to the parameter. In property get, a parameter with the same name as the property is illegal.